### PR TITLE
Add the initial chef_client_trusted_certificate resource

### PR DESCRIPTION
**Pull request by tas50**
This is the first take at a simple resource to add trusted certificates
to chef's trusted_certificate directory. This directory is not created
out of the box so we create it if it's missing with 640 perms. For the
moment the resource just takes the cert as a string. Once we build a
file fetching mixin this should allow us to specify the content as a
string or a URI.

Before using this resource fetching `https://self-signed.badssl.com/index.html` within the test cookbook resulted in the following error:

```
       * remote_file[/tmp/kitchen/cache/index.html] action create[2020-08-19T18:48:00+00:00] ERROR: SSL Validation failure connecting to host: self-signed.badssl.com - SSL_connect returned=1 errno=0 state=error: certificate verify failed (self signed certificate)


           ================================================================================
           Error executing action `create` on resource 'remote_file[/tmp/kitchen/cache/index.html]'
           ================================================================================

           OpenSSL::SSL::SSLError
           ----------------------
           SSL Error connecting to https://self-signed.badssl.com/index.html - SSL_connect returned=1 errno=0 state=error: certificate verify failed (self signed certificate)

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/test/recipes/config.rb

            34: remote_file ::File.join(Chef::Config[:file_cache_path], 'index.html') do
            35:   source 'https://self-signed.badssl.com/index.html'
            36: end
            37:

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/test/recipes/config.rb:34:in `from_file'

           remote_file("/tmp/kitchen/cache/index.html") do
             action [:create]
             default_guard_interpreter :default
             source ["https://self-signed.badssl.com/index.html"]
             declared_type :remote_file
             cookbook_name "test"
             recipe_name "config"
             remote_domain nil
             remote_user nil
             path "/tmp/kitchen/cache/index.html"
             owner nil
             group nil
             headers {}
             mode nil
             verifications []
           end

           System Info:
           ------------
           chef_version=16.4.38
           platform=centos
           platform_version=8.2.2004
           ruby=ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-linux]
           program_name=/opt/chef/bin/chef-client
           executable=/opt/chef/bin/chef-client
```

With the chef_client_trusted_cert running to add that certificate it runs without issue:

```
    * remote_file[/tmp/kitchen/cache/index.html] action create
           - create new file /tmp/kitchen/cache/index.html
           - update content in file /tmp/kitchen/cache/index.html from none to 6c8b9f
           --- /tmp/kitchen/cache/index.html	2020-08-19 18:46:46.325612910 +0000
           +++ /tmp/kitchen/cache/.chef-index20200819-6235-mr5mhi.html	2020-08-19 18:46:46.323612895 +0000
           @@ -1 +1,21 @@
           +<!DOCTYPE html>
           +<html>
           +<head>
           +  <meta charset="utf-8">
           +  <meta name="viewport" content="width=device-width, initial-scale=1">
           +  <link rel="shortcut icon" href="/icons/favicon-red.ico"/>
           +  <link rel="apple-touch-icon" href="/icons/icon-red.png"/>
           +  <title>self-signed.badssl.com</title>
           +  <link rel="stylesheet" href="/style.css">
           +  <style>body { background: red; }</style>
           +</head>
           +<body>
           +<div id="content">
           +  <h1 style="font-size: 12vw;">
           +    self-signed.<br>badssl.com
           +  </h1>
           +</div>
           +
           +</body>
           +</html>
           - restore selinux security context
```

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: resources/trusted_certificate.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+provides :chef_client_trusted_certificate
+resource_name :chef_client_trusted_certificate
+
+property :cert_name, String, name_property: true
+
+# this version check can go away once this is ported into chef itself
+property :certificate, String, required: Chef::VERSION.to_i >= 16 ? [:add] : true
+
+action :add do
+  unless ::Dir.exist?(Chef::Config[:trusted_certs_dir])
```
Comment (phiggins) [inline resources/trusted_certificate.rb:31]: I'm curious, why use a regular condition and not a `not_if`?

Comment (tas50) [reply resources/trusted_certificate.rb:31]: If you use not_if then the resource shows up in the logs each run, which I'm not a huge fan of. If you just want chef to create the cert you don't really care about the implementation details of our resource. You just want the cert to go where it needs to go and this produces a more clear log for that.

File: resources/trusted_certificate.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  file cert_path do
+    content new_resource.certificate
+    mode '0640'
+  end
+end
+
+action :remove do
+  file cert_path do
+    action :delete
+    mode '0640'
```
Comment (phiggins) [inline resources/trusted_certificate.rb:47]: Does `mode` make sense in a `:remove` action?

Comment (tas50) [reply resources/trusted_certificate.rb:47]: Nuked

**Pull request merged**